### PR TITLE
Issues/5868 sort by offset

### LINKED ISSUE #5868: Reader: Use offset instead of score for sorting search results
Title has it. 
The value of score can change so its not the most reliable method for sorting.

### PR DESCRIPTION
**Pull request by aerych**
Fixes #5868 
This also fixes a couple of issues impacting the UI.  
We were pruning some of the results (expected duplicates) to avoid doing unnecessary work but it was throwing counts off a bit. That code was removed.  
I also found that the search API sometimes returns the same post across multiple pages. When this happened the sortRank would get updated causing it to be shifted from its previous location, throwing off the placement of things in the UI. The effect was the content in the list would suddenly shift.  We no longer overwrite the original sortRank. 

To test:
Go to the reader search feature and perform a search. Scroll to invoke the infinite scroll behavior.  Confirm that the list performs as expected. 
Go to any non-search topic in the reader.  Confirm that the post continues to load as expected. 
Browse a high traffic tag like "art".  Wait about an hour or so for content to build up after your last visit.  Return to the list and let it load.  Confirm that a "gap marker" appears to let you fill in any content missing between what was just synced and what was previously synced.  Tap the gap marker and confirm that content loads correctly. 

Note:
I've noticed a couple of odd glitches with Search that I haven't been able to reproduce or guess the cause. In once instance a gap marker suddenly appeared after I'd been aggressively scrolling to load more.  In another instance, the loading spinner seemed to get "stuck".  I haven't been able to reproduce these more than once and it *might* have been related to either the previous logic.  Something we should watch for.

Needs review: @kurzee up for another one?

### CODE REVIEW

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m [hunk 1]
```diff
@@ -221,17 +223,27 @@ - (void)fetchPostsFromEndpoint:(NSURL *)endpoint
...
     [self.wordPressComRestApi GET:path
            parameters:params
               success:^(id responseObject, NSHTTPURLResponse *httpResponse) {
                   if (!success) {
                       return;
                   }
+
+                  // NOTE: If an offset param was specified sortRank will be derived
+                  // from the offset + order of the results.
+                  BOOL rankByOffset = [params objectForKey:@"offset"] != nil;
```
Comment (kurzee) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: Maybe some `const` strings for `@"offset"`.

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m [hunk 2]
```diff
@@ -9,6 +9,8 @@
 #import "WordPress-Swift.h"
 #import "PhotonImageURLHelper.h"
 
+CGFloat const ReaderPostServiceRemotOffsetRankCap = 1000000.0;
```
Comment (kurzee) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: _Maybe_ a comment here for why this number is particularly set to `1000000.0`, although it does say below that it is an arbitrarily high number.

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m [hunk 3]
```diff
@@ -221,17 +223,27 @@ - (void)fetchPostsFromEndpoint:(NSURL *)endpoint
...
            parameters:params
               success:^(id responseObject, NSHTTPURLResponse *httpResponse) {
                   if (!success) {
                       return;
                   }
+
+                  // NOTE: If an offset param was specified sortRank will be derived
+                  // from the offset + order of the results.
+                  BOOL rankByOffset = [params objectForKey:@"offset"] != nil;
+                  __block CGFloat offset = [[params numberForKey:@"offset"] floatValue];
```
Comment (kurzee) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: Maybe read and set `CGFloat offset = [[params numberForKey:@"offset"] floatValue]` in the block within `if (rankByOffset)`, as an optimization.

Comment (kurzee) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: Not sure if there would be any unexpected results if `rankByOffset` was evaluated as `true` while the actual value for `[[params numberForKey:@"offset"] floatValue]` was `0` as a valid `NSNumber` obj?

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m [hunk 4]
```diff
@@ -242,6 +254,17 @@ - (void)fetchPostsFromEndpoint:(NSURL *)endpoint
               }];
 }
 
+- (RemoteReaderPost *)formatPostDictionary:(NSDictionary *)dict offset:(CGFloat)offset
+{
+    RemoteReaderPost *post = [self formatPostDictionary:dict];
+    // Its assumed that sortRank values are in descending order. Since
```
Comment (kurzee) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: `Its` > `It's`.... suuuuuuuuper nitpicky.

File: WordPress/Classes/Services/ReaderPostService.m [hunk 5]
```diff
@@ -1015,7 +1010,16 @@ - (ReaderPost *)createOrReplaceFromRemotePost:(RemoteReaderPost *)remotePost for
     post.score = remotePost.score;
     post.siteID = remotePost.siteID;
     post.sortDate = remotePost.sortDate;
-    post.sortRank = remotePost.sortRank;
+
+    if (!existing) {
+        // Failsafe.  The `read/searc` endpoint might return the same post on
```
Comment (kurzee) [inline WordPress/Classes/Services/ReaderPostService.m]: `read/searc` > `read/search`.... _runs away_

File: WordPress/Classes/Services/ReaderPostService.m [hunk 6]
```diff
@@ -1015,7 +1010,16 @@ - (ReaderPost *)createOrReplaceFromRemotePost:(RemoteReaderPost *)remotePost for
...
+        // more than one page. If this happens preserve the *original* sortRank
+        // to avoid content jumping around in the UI.
+        // Posts from other endpoits will store a date value which shouldn't
```
Comment (kurzee) [inline WordPress/Classes/Services/ReaderPostService.m]: `endpoits` > `endpoints`... _hides in a tree_

**Comment by kurzee** [comment]
Encountered an assertion failure:

```
2016-08-16 18:36:40:255 WordPress[57522:12167683] Deleting Gap Marker: <ReaderGapMarker: 0x7ff707800320> (entity: ReaderGapMarker; id: 0xd000000011e4001e <x-coredata://A2B331EB-9F6F-486D-966C-27426166D351/ReaderGapMarker/p1145> ; data: <fault>)
fatal error: Search topics should no have a gap to fill.: file ~WordPress-iOS/WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift, line 1231
```

Steps were as follows:

1. Built and ran branch.
2. Opened Reader.
3. Searched a previously searched topic.
4. Scrolled down quickly to page.
5. Instead scrolling offset jumped up to a newly detected gap.
6. Selected to load gap.
7. Assertion failure.

screen shot 2016-08-16 at 6 39 23 pm

**Comment by kurzee** [comment]
Still experiencing a lot of jumps and unexpected loading while paging.

Here's the steps I followed:

1. Built and ran branch.
2. Opened Reader.
3. Searched a previously searched topic.
4. Scrolled down quickly to page.
5. Paging broke the offset to around the top of the tableView.
6. Paging failed to load in the actual next page of the posts.

**Comment by kurzee** [comment]
Paging seems to work as expected if I enter a new search, one that was not previously searched and having any cached posts laying around.

Perhaps the issues I posted above have to do with my previous testing in the other PR? I'll test this branch again a bit later with a topic I searched just now, perhaps it too will break having cached posts laying around and newer posts available.

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m (hunk 2 above)
Comment (aerych) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: Actually, I think I'll drop the constant completely and [adopt the approach](https://github.com/wordpress-mobile/WordPress-Android/pull/4500) @nbradbury used in the android app. Its a more elegant solution. :)

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m (hunk 3 above)
Comment (aerych) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: > Maybe read and set CGFloat offset = [[params numberForKey:@"offset"] floatValue] in the block within if (rankByOffset), as an optimization.

I don't think that would work as we need to increment the value per pass through the block. Also, having to recreate the var each pass doesn't seem like an optimization as setting it once outside the block.  But maybe I've misunderstood what you had in mind? 

> Not sure if there would be any unexpected results if rankByOffset was evaluated as true while the actual value for [[params numberForKey:@"offset"] floatValue] was 0 as a valid NSNumber obj?

Good eye. I'll add a safety net for that circumstance.

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m (hunk 4 above)
Comment (aerych) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: Hey, I'm just glad *someone* is there to keep my grammar on point! :D

File: WordPress/Classes/Services/ReaderPostService.m (hunk 6 above)
Comment (aerych) [inline WordPress/Classes/Services/ReaderPostService.m]: Can I just blame my keyboard keys?  They're broken.  Yeah.  That's what happened.

**Comment by kurzee** [comment]
After opening up this branch and researching the same search I left off on last night, I crashed while scrolling down a couple of posts (no paging):

```
2016-08-17 11:15:31.587 WordPress[66840:12377028] CoreData: warning: An NSManagedObjectContext delegate overrode fault handling behavior to silently delete the object with ID '0xd0000000138c001c <x-coredata://A2B331EB-9F6F-486D-966C-27426166D351/ReaderPost/p1251>' and substitute nil/0 for all property values instead of throwing.
2016-08-17 11:15:31.594 WordPress[66840:12377028] CoreData: warning: An NSManagedObjectContext delegate overrode fault handling behavior to silently delete the object with ID '0xd00000001390001c <x-coredata://A2B331EB-9F6F-486D-966C-27426166D351/ReaderPost/p1252>' and substitute nil/0 for all property values instead of throwing.
2016-08-17 11:15:31.600 WordPress[66840:12377028] CoreData: warning: An NSManagedObjectContext delegate overrode fault handling behavior to silently delete the object with ID '0xd00000001398001c <x-coredata://A2B331EB-9F6F-486D-966C-27426166D351/ReaderPost/p1254>' and substitute nil/0 for all property values instead of throwing.
2016-08-17 11:15:31.611 WordPress[66840:12377028] CoreData: warning: An NSManagedObjectContext delegate overrode fault handling behavior to silently delete the object with ID '0xd00000001384001c <x-coredata://A2B331EB-9F6F-486D-966C-27426166D351/ReaderPost/p1249>' and substitute nil/0 for all property values instead of throwing.
2016-08-17 11:15:31:655 WordPress[66840:12379952] Error fetching post at a specific offset.
2016-08-17 11:15:34.387 WordPress[66840:12377028] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Invalid predicate: nil RHS'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000115946d85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x00000001153badeb objc_exception_throw + 48
	2   CoreData                            0x0000000114007461 -[NSComparisonPredicate(_NSCoreDataSQLPredicateCategories) minimalFormInContext:] + 1009
	3   CoreData                            0x0000000114006c2b -[NSCompoundPredicateOperator(_NSCoreDataSQLPredicateCategories) minimalFormInContext:ofPredicates:] + 315
	4   CoreData                            0x0000000114006c2b -[NSCompoundPredicateOperator(_NSCoreDataSQLPredicateCategories) minimalFormInContext:ofPredicates:] + 315
	5   CoreData                            0x00000001141135c6 -[NSSQLGenerator initializeContextForRequest:ignoreInheritance:nestingLevel:] + 486
	6   CoreData                            0x0000000114113764 -[NSSQLGenerator initializeContextForFetchRequest:ignoreInheritance:nestingLevel:nestIsWhereScoped:] + 36
	7   CoreData                            0x0000000114113a32 -[NSSQLGenerator newSQLStatementForRequest:ignoreInheritance:countOnly:nestingLevel:nestIsWhereScoped:] + 98
	8   CoreData                            0x00000001140067e3 -[NSSQLGenerator newSQLStatementForFetchRequest:ignoreInheritance:countOnly:nestingLevel:nestIsWhereScoped:] + 35
	9   CoreData                            0x00000001140f97b8 -[NSSQLAdapter _statementForFetchRequest:ignoreInheritance:countOnly:nestingLevel:] + 344
	10  CoreData                            0x000000011400667c -[NSSQLAdapter _newSelectStatementWithFetchRequest:ignoreInheritance:] + 316
	11  CoreData                            0x00000001140062f6 -[NSSQLCore newRowsForFetchPlan:] + 118
	12  CoreData                            0x0000000114005bac -[NSSQLCore objectsForFetchRequest:inContext:] + 524
	13  CoreData                            0x0000000114005549 -[NSSQLCore executeRequest:withContext:error:] + 377
	14  CoreData                            0x00000001140eaeaf __65-[NSPersistentStoreCoordinator executeRequest:withContext:error:]_block_invoke + 3311
	15  CoreData                            0x00000001140f44dd gutsOfBlockToNSPersistentStoreCoordinatorPerform + 189
	16  libdispatch.dylib                   0x00000001167223eb _dispatch_client_callout + 8
	17  libdispatch.dylib                   0x0000000116706ef5 _dispatch_barrier_sync_f_invoke + 393
	18  CoreData                            0x00000001140e55d5 _perform + 197
	19  CoreData                            0x00000001140051b4 -[NSPersistentStoreCoordinator executeRequest:withContext:error:] + 500
	20  CoreData                            0x0000000114003973 -[NSManagedObjectContext executeFetchRequest:error:] + 579
	21  CoreData                            0x000000011404b87b -[NSManagedObjectContext(_NestedContextSupport) _parentObjectsForFetchRequest:inContext:error:] + 427
	22  CoreData                            0x00000001140c467c __82-[NSManagedObjectContext(_NestedContextSupport) executeRequest:withContext:error:]_block_invoke + 364
	23  CoreData                            0x000000011404b684 internalBlockToNSManagedObjectContextPerform + 84
	24  libdispatch.dylib                   0x00000001167223eb _dispatch_client_callout + 8
	25  libdispatch.dylib                   0x0000000116706ef5 _dispatch_barrier_sync_f_invoke + 393
	26  CoreData                            0x000000011404b58f _perform + 223
	27  CoreData                            0x000000011404b3c8 -[NSManagedObjectContext(_NestedContextSupport) executeRequest:withContext:error:] + 296
	28  CoreData                            0x0000000114003973 -[NSManagedObjectContext executeFetchRequest:error:] + 579
	29  CoreData                            0x000000011404b87b -[NSManagedObjectContext(_NestedContextSupport) _parentObjectsForFetchRequest:inContext:error:] + 427
	30  CoreData                            0x00000001140c467c __82-[NSManagedObjectContext(_NestedContextSupport) executeRequest:withContext:error:]_block_invoke + 364
	31  CoreData                            0x000000011404b684 internalBlockToNSManagedObjectContextPerform + 84
	32  libdispatch.dylib                   0x00000001167223eb _dispatch_client_callout + 8
	33  libdispatch.dylib                   0x000000011670bb7c _dispatch_barrier_sync_f_slow_invoke + 284
	34  libdispatch.dylib                   0x00000001167223eb _dispatch_client_callout + 8
	35  libdispatch.dylib                   0x000000011670a1ef _dispatch_main_queue_callback_4CF + 1738
	36  CoreFoundation                      0x00000001158a00f9 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	37  CoreFoundation                      0x0000000115861b99 __CFRunLoopRun + 2073
	38  CoreFoundation                      0x00000001158610f8 CFRunLoopRunSpecific + 488
	39  GraphicsServices                    0x000000011829aad2 GSEventRunModal + 161
	40  UIKit                               0x0000000111647f09 UIApplicationMain + 171
	41  WordPress                           0x000000010dcbbaae main + 62
	42  libdyld.dylib                       0x000000011675692d start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by kurzee** [comment]
Revisiting this branch and searching another topic that I searched last night presented some very buggy paging and offset jumps while loading. I saw several gaps appears in the list but they would also quickly disappear, I suppose while other paging requests were finishing and updating the results list.

I'm thinking it may not be feasible to support cacheing search results that are likely to have gaps, and maybe instead start fresh while searching, or at least on the next app session.

File: WordPress/Classes/Networking/ReaderPostServiceRemote.m (hunk 3 above)
Comment (kurzee) [inline WordPress/Classes/Networking/ReaderPostServiceRemote.m]: >I don't think that would work as we need to increment the value per pass through the block. Also, having to recreate the var each pass doesn't seem like an optimization as setting it once outside the block. But maybe I've misunderstood what you had in mind?

Ah my fault, I overlooked what was going on.

**Comment by aerych** [comment]
Thanks for some amazing testing @kurzee. 

Just to reiterate here, I've tracked a few of the glitches to some quirks in our API and I attribute at least some of the jumpiness there.  That's something I can work with the squad to address server-side. 

The issue with the stuck loading spinner was traced back to 503 errors returned by the API due to throttling and those errors not being properly handled.  Now the error is handled and when it occurs the "end of list" table footer will be visible.  You can still juggle the scroll to re-trigger the load more and the spinner will appear.  Useful for getting more results after the throttling expires.

This guy...
``` 2016-08-17 11:15:31.587 WordPress[66840:12377028] CoreData: warning: An NSManagedObjectContext delegate overrode fault handling behavior to silently delete the object with ID '0xd0000000138c001c <x-coredata://A2B331EB-9F6F-486D-966C-27426166D351/ReaderPost/p1251>' and substitute nil/0 for all property values instead of throwing.```
I spent a while chasing this rabbit.  Its caused by returning to a search with existing results and quickly scrolling down while its loading. When the initial batch of search results is loaded older, stale results were being deleted.  This was a problem because they were already visible in the list.  Hence the errors and in some situations crashes.  This was a tricky one to get sorted and I was rather stumped until... 

> may not be feasible to support cacheing search

This was bugging me and it took me a while to realize why. By design searches were intended to be ephemeral and we shouldn't be caching.  It finally clicked that they were actually being persisted when they shouldn't be.  The old `ReaderViewController` contained logic for cleaning up non-menu topics and orphaned posts. This was something that should have been migrated to the new `ReaderMenuViewController` but was overlooked.  I've restored this behavior and now search topics are cleaned up when the app is launched (unless they are needed for state restoration).  I've added some additional logic to clean up topics when the search controller is dismissed, and when searching a new topic.  This should address the core data warnings we were seeing. 

Things *seem* solid to me... or at least I can't reproduce any of the above issues.  

Ready for another round at your leisure.

**Comment by kurzee** [comment]
Every now and then, there seems to be a main thread lockup while paging. I can't pin exact steps to reproduce, but something along the lines of:

1. Search a topic.
2. Scroll down to load the next page.
3. _Sometimes_ the UI will lock, with a spinning activity indicator, but no other interaction available.
4. Eventually the UI will unlock and another page will load or finishes as the throttled end of the list.

**Comment by kurzee** [comment]
If state restoration restores the original results of a search, assuming they left off there, should that list be purged/refreshed just after the state restores?

I wonder because a far outdated state-restored list might present the same issues as with "cacheing posts" while paging. Maybe not?

**Comment by kurzee** [comment]
@aerych other than what I mentioned above, the code looks good and it's much more solid now! I have reproduced the UI-blocking issue a couple times, just by continually paging.... If you wanted to create a separate issue/PR for that, fine by me!

**Comment by aerych** [comment]
> Every now and then, there seems to be a main thread lockup while paging

This looks like its a bug due to certain specific content and unrelated to the changes here Excellent luck finding it tho! I've opened an [issue to fix it](https://github.com/wordpress-mobile/WordPress-iOS/issues/5878).

> If state restoration restores the original results of a search, assuming they left off there, should that list be purged/refreshed just after the state restores?

Probably not. I'd be worried that the user had left off on the post detail page, and purging/refreshing isn't guaranteed to restore a particular post.

> I wonder because a far outdated state-restored list might present the same issues as with "cacheing posts" while paging. Maybe not?

Most likely not but I've added an additional safety net to ensure that we do not perform the initial fetch of a search topic's content if that topic already has content loaded. I think we should be good that way.

**Comment by kurzee** [comment]
Awesome work! :shipit:!

**Comment by aerych** [comment]
Amazing review sir. Many thanks!!

**Pull request merged**